File: frameworks/language/benchmarkdotnet/SampleBenchmark.cs
// BenchmarkDotNet implementation of the canonical benchzoo sample
// benchmark. See ../../../docs/sample-benchmark.md for the spec.
//
// A few BenchmarkDotNet-specific notes that drive the attribute choices
// below:
//
// * BenchmarkDotNet's default job runs a multi-stage pilot + warmup +
//   many-iteration main phase, which is wildly inappropriate for a
//   benchmark whose headline measurement is a 2.15-second Thread.Sleep.
//   A default run of Benchmark1 would sleep for minutes. We use
//   [SimpleJob(RunStrategy.Monitoring, ...)] — Monitoring is the
//   strategy BenchmarkDotNet documents for long / noisy / real-world
//   measurements where per-invocation overhead is negligible and you
//   just want a small fixed number of actual invocations measured.
//   iterationCount: 3 keeps total wall time modest (roughly
//   3 * (2.15 + small + small + 2.15) seconds plus startup) while still
//   producing enough samples for Statistics.Mean / StandardDeviation to
//   be meaningful.
//
// * [JsonExporterAttribute.Full] makes BenchmarkDotNet emit the full
//   JSON report (the "-report-full.json" variant) which includes both
//   the Statistics dict and the per-iteration Measurements[] array.
//   That's the format run.sh copies to ./output.json for the workflow
//   artifact. [CsvExporter] additionally emits a CSV report with
//   columns like Method, Job, Mean, Error, StdDev, etc., which run.sh
//   copies to ./output.csv.
//
// * Benchmark2 returns the loop accumulator. BenchmarkDotNet's user
//   guide explicitly documents that returning a value from a benchmark
//   method is the idiomatic way to prevent dead-code elimination: the
//   runtime can't prove the returned value is unused, so the JIT keeps
//   the computation. The other three tests don't need this guard
//   (Thread.Sleep, file I/O, and DateTime.UtcNow are all side-effecting
//   and cannot be optimised away).
//
// * Benchmark3 writes to /dev/null. That path exists only on Unix, so
//   this class will throw on Windows. The workflow pins
//   runs-on: ubuntu-latest, which is fine, but running this under
//   Windows or macOS without adjustment will fail at Benchmark3.

using System;
using System.IO;
using System.Threading;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Engines;
using BenchmarkDotNet.Exporters.Csv;
using BenchmarkDotNet.Jobs;

namespace BenchzooSample;

[JsonExporterAttribute.Full]
[CsvExporter]
[SimpleJob(RunStrategy.Monitoring, iterationCount: 3)]
public class SampleBenchmark
{
    // Test 1 — sleep-dominated (~2.15 s). Void return is fine here:
    // Thread.Sleep is a side-effecting syscall, so there is nothing for
    // the JIT to elide.
    [Benchmark]
    public void Benchmark1()
    {
        Thread.Sleep(2150);
    }

    // Test 2 — tight CPU loop (sub-millisecond). Returning the
    // accumulator is the BenchmarkDotNet-documented idiom for
    // suppressing dead-code elimination; without it the JIT would be
    // free to delete the loop body entirely and report zero.
    [Benchmark]
    public int Benchmark2()
    {
        int sum = 0;
        for (int i = 0; i < 1000; i++)
        {
            sum += i;
        }
        return sum;
    }

    // Test 3 — write 1,400,000 bytes of pseudo-random data to
    // /dev/null. The byte buffer is filled from a seeded Random so the
    // output is deterministic across runs. The /dev/null path is
    // Linux-specific (and macOS-compatible); on Windows this throws.
    // The workflow pins ubuntu-latest, matching that assumption.
    [Benchmark]
    public void Benchmark3()
    {
        var buffer = new byte[1_400_000];
        var rng = new Random(seed: 42);
        rng.NextBytes(buffer);
        using var stream = File.OpenWrite("/dev/null");
        stream.Write(buffer, 0, buffer.Length);
    }

    // Test 4 — monthly change-point showcase. sleep_time = 2.15 +
    // ((m mod 3) - 1) where m is the current month in UTC. Using
    // DateTime.UtcNow (not DateTime.Now) is load-bearing: the test is
    // supposed to produce the same value on the same calendar day
    // regardless of runner timezone. See docs/sample-benchmark.md.
    [Benchmark]
    public void Benchmark4()
    {
        int month = DateTime.UtcNow.Month;
        double sleepSeconds = 2.15 + ((month % 3) - 1);
        Thread.Sleep((int)(sleepSeconds * 1000));
    }
}

File: frameworks/language/jmh/src/main/java/io/nyrkio/benchzoo/jmh/SampleBenchmark.java
package io.nyrkio.benchzoo.jmh;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.Warmup;
import org.openjdk.jmh.infra.Blackhole;

import java.io.FileOutputStream;
import java.io.IOException;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.Random;
import java.util.concurrent.TimeUnit;

/**
 * JMH implementation of the canonical benchzoo sample benchmark.
 *
 * <p>See {@code docs/sample-benchmark.md} for the spec. Four @Benchmark
 * methods — {@code benchmark1}..{@code benchmark4} — mirror the four
 * canonical tests. JMH emits one entry per @Benchmark method in its JSON
 * output, keyed by the fully-qualified name
 * {@code io.nyrkio.benchzoo.jmh.SampleBenchmark.benchmarkN}; the parser
 * is expected to strip the class prefix and use the short
 * {@code benchmarkN} as {@code attributes["test_name"]}.
 *
 * <h2>Run-time trade-off</h2>
 *
 * Class-level {@code @Warmup(iterations = 1, time = 1)},
 * {@code @Measurement(iterations = 3, time = 1)}, and {@code @Fork(1)}
 * are deliberately modest so the whole suite finishes in a few minutes
 * of CI wall time.
 *
 * <p>JMH's {@code time} attribute on {@code @Measurement}/{@code @Warmup}
 * is the <em>target batch duration</em>: JMH repeatedly invokes the
 * benchmark method until roughly {@code time} seconds have elapsed, then
 * treats that as one measurement iteration. For the sub-millisecond
 * {@code benchmark2} this produces millions of invocations per iteration,
 * which is exactly what you want for statistical stability. For the
 * sleep-heavy {@code benchmark1} and {@code benchmark4} this is the
 * <em>opposite</em> of what you want: a ~2.15s sleep easily overshoots
 * the 1s target, JMH reacts by running a single invocation per
 * iteration, and the measurement iteration still takes ~2.15 s. Over 3
 * iterations × 1 fork that's ~6.5 s per sleep benchmark, plus 1 warmup
 * iteration. Total suite wall time is on the order of a minute, which
 * is acceptable for CI.
 *
 * <p>We considered overriding {@code @Measurement(iterations = 1)} per
 * sleep method to shave seconds off the run, but chose to keep a uniform
 * 3-iteration measurement across all four benchmarks: the fixture stays
 * easier to reason about (same {@code measurementIterations} in every
 * JSON entry), and the ground-truth assertion on {@code benchmark1}'s
 * mean is slightly more robust with three samples than with one.
 */
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@Warmup(iterations = 1, time = 1)
@Measurement(iterations = 3, time = 1)
@Fork(1)
@State(Scope.Benchmark)
public class SampleBenchmark {

    /**
     * Seeded RNG for benchmark3, to keep the random payload deterministic
     * run-to-run. The actual byte values do not matter for wall-clock
     * measurement, but a fixed seed makes differential debugging easier.
     */
    private final Random random = new Random(42);

    /**
     * Pre-allocated 1,400,000-byte scratch buffer for benchmark3. We fill
     * it fresh on each invocation via {@link Random#nextBytes(byte[])},
     * so the "random data" is regenerated per call but the allocation
     * cost is kept out of the measured path.
     */
    private final byte[] payload = new byte[1_400_000];

    /**
     * Test 1 — sleep-dominated (~2.15 s).
     *
     * <p>{@code Thread.sleep(2150)} throws {@link InterruptedException},
     * which JMH allows benchmark methods to declare.
     */
    @Benchmark
    public void benchmark1() throws InterruptedException {
        Thread.sleep(2150);
    }

    /**
     * Test 2 — tight CPU loop, sub-millisecond.
     *
     * <p>{@link Blackhole#consume(int)} is <strong>mandatory</strong>
     * here. Without it, HotSpot's JIT trivially proves the loop has no
     * observable side effects and eliminates it entirely, which would
     * make this benchmark measure "return from an empty method" rather
     * than "execute 1000 loop iterations." The Blackhole API exists
     * specifically to defeat dead-code elimination in JMH benchmarks;
     * see the JMH samples for the canonical rationale.
     */
    @Benchmark
    public void benchmark2(Blackhole blackhole) {
        for (int i = 0; i < 1000; i++) {
            blackhole.consume(i);
        }
    }

    /**
     * Test 3 — write 1.4 MB of pseudo-random data to /dev/null.
     *
     * <p>Uses a {@link FileOutputStream} opened on {@code /dev/null}.
     * The buffer is filled via {@link Random#nextBytes(byte[])} inside
     * the measured path so the RNG work is part of the benchmark (same
     * as the other language implementations, where generating the
     * random bytes is not separated from writing them). The stream is
     * closed in a try-with-resources so we don't leak file descriptors
     * across JMH invocations.
     */
    @Benchmark
    public void benchmark3(Blackhole blackhole) throws IOException {
        random.nextBytes(payload);
        try (FileOutputStream out = new FileOutputStream("/dev/null")) {
            out.write(payload);
        }
        blackhole.consume(payload);
    }

    /**
     * Test 4 — monthly change-point showcase. Sleeps for
     * {@code 2.15 + ((month mod 3) - 1)} seconds, where {@code month}
     * is the current UTC month. See
     * {@code docs/sample-benchmark.md} for the formula table.
     */
    @Benchmark
    public void benchmark4() throws InterruptedException {
        int month = ZonedDateTime.now(ZoneOffset.UTC).getMonthValue();
        long sleepMillis = (long) ((2.15 + (month % 3 - 1)) * 1000);
        Thread.sleep(sleepMillis);
    }
}
